code/modules/holomap/mapper.dm

```diff
@@ -0,0 +1,446 @@
+				var/datum/holomap_marker/holomarker = holomap_markers[marker]
+				if(holomarker.z == T_z && holomarker.filter & mapper_filter)
+					var/image/markerImage = image(holomarker.icon,holomarker.id)
+					markerImage.plane = FLOAT_PLANE
+					markerImage.layer = FLOAT_LAYER
+					markerImage.appearance_flags = RESET_COLOR|PIXEL_SCALE
+					markerImage.pixel_x = holomarker.x+holomarker.offset_x
+					markerImage.pixel_y = holomarker.y+holomarker.offset_y
+					map_app.overlays += markerImage
+
+			var/obj/screen/mapper/map/tmp = new()
+			tmp.appearance = map_app
+			map_image_cache[map_cache_key] = tmp
+
+	bgmap = map_image_cache[map_cache_key]
+
+	// The holomap moves around, the user is always in the center. This slides the holomap.
+	var/offset_x = bgmap.offset_x
+	var/offset_y = bgmap.offset_y
+	extras_holder.pixel_x = bgmap.pixel_x = -1*T_x + offset_x
+	extras_holder.pixel_y = bgmap.pixel_y = -1*T_y + offset_y
+
+	// Populate other mapper icons
+	for(var/hc in mapping_units)
+		var/obj/item/device/mapping_unit/HC = hc
+		if(HC.mapper_filter != mapper_filter)
+			continue
+		var/mob_indicator = HOLOMAP_ERROR
+		var/turf/TU = get_turf(HC)
+		if(TU.z != T_z)
+			continue
+		
+		// Marker not on a turf
+		if(!TU)
+			continue
+		
+		// We're the marker
+		if(HC == src)
+			mob_indicator = HOLOMAP_YOU
+		
+		// The marker is held by a borg
+		else if(isrobot(HC.loc))
+			var/mob/living/silicon/robot/R = HC.loc
+			if(R.stat == DEAD)
+				mob_indicator = HOLOMAP_DEAD
+			else
+				mob_indicator = HOLOMAP_OTHER
+		
+		// The marker is worn by a human
+		else if(ishuman(loc))
+			var/mob/living/carbon/human/H = loc
+			if(H.stat == DEAD)
+				mob_indicator = HOLOMAP_DEAD
+			else
+				mob_indicator = HOLOMAP_OTHER
+		
+		// It's not attached to anything useful
+		else
+			mob_indicator = HOLOMAP_DEAD
+		
+		// Ask it to update it's icon based on helmet (or whatever)
+		HC.update_marker()
+
+		// Generate the icon and apply it to the list of images to show the client
+		if(mob_indicator != HOLOMAP_ERROR)
+
+			// This is so specific because the icons are pixel_x and pixel_y only relative to OUR view of where THEY are relative to us
+			var/marker_cache_key = "\ref[HC]_[HC.marker_prefix]_[mob_indicator]"
+
+			if(!(marker_cache_key in icon_image_cache))
+				var/obj/screen/mapper/marker/mark = new()
+				mark.icon_state = "[HC.marker_prefix][mob_indicator]"
+				icon_image_cache[marker_cache_key] = mark
+				switch(mob_indicator)
+					if(HOLOMAP_YOU)
+						mark.layer = 3 // Above the other markers
+					if(HOLOMAP_DEAD)
+						mark.layer = 2
+					else
+						mark.layer = 1
+
+			var/obj/screen/mapper/marker/mark = icon_image_cache[marker_cache_key]
+			handle_marker(mark,TU.x,TU.y)
+			extras += mark
+
+		// Marker beacon items
+		for(var/hb in mapping_beacons)
+			var/obj/item/device/holomap_beacon/HB = hb
+			if(HB.mapper_filter != mapper_filter)
+				continue
+			
+			var/turf/TB = get_turf(HB)
+			if(TU.z != T_z)
+				continue
```

Comment:
Do we need a nullcheck for the turf here too?

Reply:
This whole section was indented wrong, it should have been TB. not TU., fixed